src/WinRT.Runtime/ActivationFactory.cs

```diff
@@ -0,0 +1,272 @@
+                {
+                    int hr = _GetActivationFactory(MarshalString.GetAbi(ref __runtimeClassId), &instancePtr);
+                    if (hr == 0)
+                    {
+                        var objRef = ObjectReference<IActivationFactoryVftbl>.Attach(ref instancePtr);
+                        return (objRef, hr);
+                    }
+                    else
+                    {
+                        return (null, hr);
+                    }
+                }
+            }
+            finally
+            {
+                MarshalInspectable<object>.DisposeAbi(instancePtr);
+            }
+        }
+
+        ~DllModule()
+        {
+            System.Diagnostics.Debug.Assert(_CanUnloadNow == null || _CanUnloadNow() == 0); // S_OK
+            lock (_cache)
+            {
+                _cache.Remove(_fileName);
+            }
+            if ((_moduleHandle != IntPtr.Zero) && !Platform.FreeLibrary(_moduleHandle))
+            {
+                Marshal.ThrowExceptionForHR(Marshal.GetHRForLastWin32Error());
+            }
+        }
+    }
+
+    internal sealed class WinrtModule
+    {
+        readonly IntPtr _mtaCookie;
+        volatile static WinrtModule _instance;
+        private static WinrtModule MakeWinRTModule()
+        {
+            global::System.Threading.Interlocked.CompareExchange(ref _instance, new WinrtModule(), null);
+            return _instance;
+        }
+        public static WinrtModule Instance => _instance ?? MakeWinRTModule();
+
+        public unsafe WinrtModule()
+        {
+            IntPtr mtaCookie;
+            Marshal.ThrowExceptionForHR(Platform.CoIncrementMTAUsage(&mtaCookie));
+            _mtaCookie = mtaCookie;
+        }
+
+        public static unsafe (ObjectReference<I> obj, int hr) GetActivationFactory<I>(string runtimeClassId, Guid iid)
+        {
+            var module = Instance; // Ensure COM is initialized
+            IntPtr instancePtr = IntPtr.Zero;
+            try
+            {
+                MarshalString.Pinnable __runtimeClassId = new(runtimeClassId);
+                fixed (void* ___runtimeClassId = __runtimeClassId)
+                {
+                    int hr = Platform.RoGetActivationFactory(MarshalString.GetAbi(ref __runtimeClassId), &iid, &instancePtr);
+                    if (hr == 0)
+                    {
+                        var objRef = ObjectReference<I>.Attach(ref instancePtr);
+                        return (objRef, hr);
+                    }
+                    else
+                    {
+                        return (null, hr);
+                    }
+                }
+            }
+            finally
+            {
+                MarshalInspectable<object>.DisposeAbi(instancePtr);
+            }
+        }
+
+        ~WinrtModule()
+        {
+            Marshal.ThrowExceptionForHR(Platform.CoDecrementMTAUsage(_mtaCookie));
+        }
+    }
+
+#if EMBED
+    internal

```

Comment:
Perhaps move this to a standalone file, so we can better separate it? Since this type is public and other two are not

Reply:
They are very much related. Not much benefit to splitting I'd say.